// 0.1.unity-settings-daemon/lib/front-core.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require("@angular/core");
var platform_browser_1 = require("@angular/platform-browser");
var forms_1 = require("@angular/forms");
var material_1 = require("@angular/material");
var login_module_1 = require("./login/login.module");
var document_viewer_module_1 = require("./document-viewer/document-viewer.module");
var results_module_1 = require("./results/results.module");
var services_1 = require("./services");
var ATLASENSE_MODULES = [
    results_module_1.ResultsModule,
    login_module_1.LoginModule,
    document_viewer_module_1.DocumentViewerModule
];
var FrontCoreRootModule = (function () {
    function FrontCoreRootModule() {
    }
    return FrontCoreRootModule;
}());
FrontCoreRootModule = __decorate([
    core_1.NgModule({
        imports: [
            platform_browser_1.BrowserModule,
            forms_1.FormsModule,
            material_1.MaterialModule.forRoot(),
            login_module_1.LoginModule.forRoot(),
            results_module_1.ResultsModule,
            document_viewer_module_1.DocumentViewerModule
        ],
        exports: ATLASENSE_MODULES,
        providers: services_1.ATLASENSE_PROVIDERS
    }),
    __metadata("design:paramtypes", [])
], FrontCoreRootModule);
exports.FrontCoreRootModule = FrontCoreRootModule;
var FrontCoreModule = (function () {
    function FrontCoreModule() {
    }
    FrontCoreModule.forRoot = function (config) {
        return {
            ngModule: FrontCoreRootModule,
            providers: [
                {
                    provide: 'api.config',
                    useValue: config
                },
                services_1.ATLASENSE_PROVIDERS
            ]
        };
    };
    return FrontCoreModule;
}());
FrontCoreModule = __decorate([
    core_1.NgModule({
        imports: ATLASENSE_MODULES,
        exports: ATLASENSE_MODULES,
    }),
    __metadata("design:paramtypes", [])
], FrontCoreModule);
exports.FrontCoreModule = FrontCoreModule;
