// 0.1.unity-settings-daemon/lib/websocket.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require("@angular/core");
var Observable_1 = require("rxjs/Observable");
var Subject_1 = require("rxjs/Subject");
var api_service_1 = require("./api.service");
var auth_service_1 = require("./auth.service");
var config_service_1 = require("./config.service");
var WEBSOCKET_URL = '@ws';
var WEBSOCKET_TOKEN_URL = '@wstoken';
var $WebSocket = $WebSocket_1 = (function () {
    function $WebSocket(api, auth, configuration) {
        var _this = this;
        this.api = api;
        this.auth = auth;
        this.configuration = configuration;
        this.reconnectAttempts = 0;
        this.sendQueue = [];
        this.onOpenCallbacks = [];
        this.onMessageCallbacks = [];
        this.onErrorCallbacks = [];
        this.onCloseCallbacks = [];
        this.readyStateConstants = {
            'CONNECTING': 0,
            'OPEN': 1,
            'CLOSING': 2,
            'CLOSED': 3,
            'RECONNECT_ABORTED': 4
        };
        this.normalCloseCode = 1000;
        this.reconnectableStatusCodes = [4000];
        this.config = { initialTimeout: 500, maxTimeout: 300000, reconnectIfNotNormalClose: false };
        this.dataStream = new Subject_1.Subject();
        this.auth.loggedin.subscribe(function (data) { return _this.preconnect(); });
    }
    $WebSocket.prototype.preconnect = function (force) {
        var _this = this;
        if (force === void 0) { force = false; }
        var endpoint = this.configuration.internal_urls.canonical;
        endpoint = endpoint + '/' + WEBSOCKET_TOKEN_URL;
        this.api.get(endpoint)
            .subscribe(function (res) { return _this.connect(res.json(), force); }, function (err) { return console.log(err); });
    };
    $WebSocket.prototype.connect = function (token, force) {
        var _this = this;
        if (force === void 0) { force = false; }
        var endpoint = this.configuration.internal_urls.canonical;
        endpoint = endpoint + '/' + WEBSOCKET_URL + '?ws_token=' + token.token;
        var self = this;
        if (force || !this.socket || this.socket.readyState !== this.readyStateConstants.OPEN) {
            self.socket = new WebSocket(endpoint);
            self.socket.onopen = function (ev) {
                _this.onOpenHandler(ev);
            };
            self.socket.onmessage = function (ev) {
                self.onMessageHandler(ev);
                _this.dataStream.next(ev);
            };
            this.socket.onclose = function (ev) {
                self.onCloseHandler(ev);
            };
            this.socket.onerror = function (ev) {
                self.onErrorHandler(ev);
                _this.dataStream.error(ev);
            };
        }
    };
    $WebSocket.prototype.send = function (data) {
        var self = this;
        if (this.getReadyState() !== this.readyStateConstants.OPEN
            && this.getReadyState() !== this.readyStateConstants.CONNECTING) {
            this.preconnect();
        }
        return Observable_1.Observable.create(function (observer) {
            if (self.socket.readyState === self.readyStateConstants.RECONNECT_ABORTED) {
                observer.next('Socket connection has been closed');
            }
            else {
                self.sendQueue.push({ message: data });
                self.fireQueue();
            }
        });
    };
    ;
    $WebSocket.prototype.getDataStream = function () {
        return this.dataStream;
    };
    $WebSocket.prototype.onOpenHandler = function (event) {
        this.reconnectAttempts = 0;
        this.notifyOpenCallbacks(event);
        this.fireQueue();
    };
    ;
    $WebSocket.prototype.notifyOpenCallbacks = function (event) {
        for (var i = 0; i < this.onOpenCallbacks.length; i++) {
            this.onOpenCallbacks[i].call(this, event);
        }
    };
    $WebSocket.prototype.fireQueue = function () {
        while (this.sendQueue.length && this.socket.readyState === this.readyStateConstants.OPEN) {
            var data = this.sendQueue.shift();
            this.socket.send($WebSocket_1.Helpers.isString(data.message) ? data.message : JSON.stringify(data.message));
        }
    };
    $WebSocket.prototype.notifyCloseCallbacks = function (event) {
        for (var i = 0; i < this.onCloseCallbacks.length; i++) {
            this.onCloseCallbacks[i].call(this, event);
        }
    };
    $WebSocket.prototype.notifyErrorCallbacks = function (event) {
        for (var i = 0; i < this.onErrorCallbacks.length; i++) {
            this.onErrorCallbacks[i].call(this, event);
        }
    };
    $WebSocket.prototype.onOpen = function (cb) {
        this.onOpenCallbacks.push(cb);
        return this;
    };
    ;
    $WebSocket.prototype.onClose = function (cb) {
        this.onCloseCallbacks.push(cb);
        return this;
    };
    $WebSocket.prototype.onError = function (cb) {
        this.onErrorCallbacks.push(cb);
        return this;
    };
    ;
    $WebSocket.prototype.onMessage = function (callback, options) {
        if (!$WebSocket_1.Helpers.isFunction(callback)) {
            throw new Error('Callback must be a function');
        }
        this.onMessageCallbacks.push({
            fn: callback,
            pattern: options ? options.filter : undefined,
            autoApply: options ? options.autoApply : true
        });
        return this;
    };
    $WebSocket.prototype.onMessageHandler = function (message) {
        var self = this;
        var currentCallback;
        for (var i = 0; i < self.onMessageCallbacks.length; i++) {
            currentCallback = self.onMessageCallbacks[i];
            currentCallback.fn.apply(self, [message]);
        }
    };
    ;
    $WebSocket.prototype.onCloseHandler = function (event) {
        this.notifyCloseCallbacks(event);
        if ((this.config.reconnectIfNotNormalClose && event.code !== this.normalCloseCode)
            || this.reconnectableStatusCodes.indexOf(event.code) > -1) {
            this.reconnect();
        }
        else {
            this.dataStream.complete();
        }
    };
    ;
    $WebSocket.prototype.onErrorHandler = function (event) {
        this.notifyErrorCallbacks(event);
    };
    ;
    $WebSocket.prototype.reconnect = function () {
        var _this = this;
        this.close(true);
        var backoffDelay = this.getBackoffDelay(++this.reconnectAttempts);
        setTimeout(function () { return _this.preconnect(); }, backoffDelay);
        return this;
    };
    $WebSocket.prototype.close = function (force) {
        if (force || !this.socket.bufferedAmount) {
            this.socket.close();
        }
        return this;
    };
    ;
    $WebSocket.prototype.getBackoffDelay = function (attempt) {
        var R = Math.random() + 1;
        var T = this.config.initialTimeout;
        var F = 2;
        var N = attempt;
        var M = this.config.maxTimeout;
        return Math.floor(Math.min(R * T * Math.pow(F, N), M));
    };
    ;
    $WebSocket.prototype.setInternalState = function (state) {
        if (Math.floor(state) !== state || state < 0 || state > 4) {
            throw new Error('state must be an integer between 0 and 4, got: ' + state);
        }
        this.internalConnectionState = state;
    };
    $WebSocket.prototype.getReadyState = function () {
        if (this.socket == null) {
            return -1;
        }
        return this.internalConnectionState || this.socket.readyState;
    };
    return $WebSocket;
}());
$WebSocket.Helpers = (function () {
    function class_1() {
    }
    class_1.isPresent = function (obj) {
        return obj !== undefined && obj !== null;
    };
    class_1.isString = function (obj) {
        return typeof obj === 'string';
    };
    class_1.isArray = function (obj) {
        return Array.isArray(obj);
    };
    class_1.isFunction = function (obj) {
        return typeof obj === 'function';
    };
    return class_1;
}());
$WebSocket = $WebSocket_1 = __decorate([
    core_1.Injectable(),
    __metadata("design:paramtypes", [api_service_1.AtlasenseapiService,
        auth_service_1.AuthService,
        config_service_1.ConfigurationService])
], $WebSocket);
exports.$WebSocket = $WebSocket;
var $WebSocket_1;
