// 0.1.unity-settings-daemon/lib/document-viewer/document-viewer.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require("@angular/core");
var document_1 = require("../document");
var DocumentViewerComponent = (function () {
    function DocumentViewerComponent() {
        this.showInfo = true;
        this.displayInfoPanel = 'show-class';
        this.displayInfoMinPanel = 'hide-class';
        this.showMoreMetaData = false;
        this.moreMetaDataClass = 'hide-class';
        this.toggleMetaData = "Show More";
    }
    DocumentViewerComponent.prototype.toggleInfo = function () {
        this.showInfo = !this.showInfo;
        this.displayInfoPanel = this.showInfo ? 'show-class' : 'hide-class';
        this.displayInfoMinPanel = this.showInfo ? 'hide-class' : 'show-class';
    };
    ;
    DocumentViewerComponent.prototype.toggleMoreMetaData = function () {
        this.showMoreMetaData = !this.showMoreMetaData;
        this.moreMetaDataClass = this.showMoreMetaData ? 'show-class' : 'hide-class';
        this.toggleMetaData = this.showMoreMetaData ? 'Show More' : 'Show Less';
    };
    DocumentViewerComponent.prototype.closeViewer = function () {
        this.document = undefined;
    };
    DocumentViewerComponent.prototype.ngOnInit = function () {
    };
    return DocumentViewerComponent;
}());
__decorate([
    core_1.Input(),
    __metadata("design:type", document_1.ASDocument)
], DocumentViewerComponent.prototype, "document", void 0);
DocumentViewerComponent = __decorate([
    core_1.Component({
        selector: 'document-viewer',
        templateUrl: './document-viewer.component.html',
        styleUrls: ['./document-viewer.component.css'],
    }),
    __metadata("design:paramtypes", [])
], DocumentViewerComponent);
exports.DocumentViewerComponent = DocumentViewerComponent;
