// 0.1.unity-settings-daemon/lib/context-menu/context-menu.item.directive.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require("@angular/core");
var ContextMenuItemDirective = (function () {
    function ContextMenuItemDirective(template) {
        this.template = template;
        this.divider = false;
        this.enabled = true;
        this.visible = true;
        this.execute = new core_1.EventEmitter();
    }
    ContextMenuItemDirective.prototype.evaluateIfFunction = function (value, item) {
        if (value instanceof Function) {
            return value(item);
        }
        return value;
    };
    ContextMenuItemDirective.prototype.triggerExecute = function (item, $event) {
        if (!this.evaluateIfFunction(this.enabled, item)) {
            return;
        }
        this.execute.emit({ event: $event, item: item });
    };
    return ContextMenuItemDirective;
}());
__decorate([
    core_1.Input(),
    __metadata("design:type", Boolean)
], ContextMenuItemDirective.prototype, "divider", void 0);
__decorate([
    core_1.Input(),
    __metadata("design:type", Object)
], ContextMenuItemDirective.prototype, "enabled", void 0);
__decorate([
    core_1.Input(),
    __metadata("design:type", Object)
], ContextMenuItemDirective.prototype, "visible", void 0);
__decorate([
    core_1.Output(),
    __metadata("design:type", core_1.EventEmitter)
], ContextMenuItemDirective.prototype, "execute", void 0);
ContextMenuItemDirective = __decorate([
    core_1.Directive({
        selector: 'template[contextMenuItem]',
    }),
    __metadata("design:paramtypes", [core_1.TemplateRef])
], ContextMenuItemDirective);
exports.ContextMenuItemDirective = ContextMenuItemDirective;
