// 0.1.unity-settings-daemon/lib/user/user.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require("@angular/core");
var api_service_1 = require("../api.service");
var config_service_1 = require("../config.service");
var USER_ENDPOINT = 'user';
var UserService = (function () {
    function UserService(api, config) {
        this.api = api;
        this.config = config;
    }
    UserService.prototype.getUser = function (user_id) {
        var user = {
            avatar: "assets/avatar.png",
            name: "Anna Atkinson",
            email: "anna@organization.com",
            data_sources: [
                "datasource1",
                "datasource2",
                "datasource3",
                "datasource4",
                "datasource5"
            ],
            role: "owner",
            blocked: true
        };
        return user;
    };
    UserService.prototype.getUsers = function () {
        var users = [
            {
                avatar: "assets/avatar.png",
                name: "Anna Atkinson",
                email: "anna@organization.com",
                data_sources: [
                    "datasource1",
                    "datasource2",
                    "datasource3",
                    "datasource4",
                    "datasource5"
                ],
                role: "owner",
                blocked: true
            },
            {
                avatar: "assets/avatar.png",
                name: "Brian Carr",
                email: "brian@organization.com",
                data_sources: [
                    "datasource1",
                    "datasource3",
                    "datasource4",
                    "datasource5"
                ],
                role: "owner",
                blocked: false
            },
            {
                avatar: "assets/avatar.png",
                name: "Jesse Li",
                email: "jesse@organization.com",
                data_sources: [
                    "datasource3",
                    "datasource5"
                ],
                role: "collaborator",
                blocked: false
            },
            {
                avatar: "assets/avatar.png",
                name: "Sasha Nikolaev",
                email: "sasha@organization.com",
                data_sources: [],
                role: "reader",
                blocked: false
            }
        ];
        return users;
    };
    return UserService;
}());
UserService = __decorate([
    core_1.Injectable(),
    __metadata("design:paramtypes", [api_service_1.AtlasenseapiService,
        config_service_1.ConfigurationService])
], UserService);
exports.UserService = UserService;
